curl/inc/leatherman/curl/client.hpp

```diff
@@ -0,0 +1,217 @@
+/**
+* @file
+* Declares the HTTP client.
+*/
+#pragma once
+
+#include <leatherman/util/scoped_resource.hpp>
+#include "request.hpp"
+#include "response.hpp"
+#include <curl/curl.h>
+
+
+namespace leatherman { namespace curl {
+
+    /**
+     * Resource for a cURL handle.
+     */
+    struct curl_handle : util::scoped_resource<CURL*>
+    {
+        /**
+         * Constructs a cURL handle.
+         */
+        curl_handle();
+
+     private:
+        static void cleanup(CURL* curl);
+    };
+
+    /**
+     * Resource for a cURL linked-list.
+     */
+    struct curl_list : util::scoped_resource<curl_slist*>
+    {
+        /**
+         * Constructs a curl_list.
+         */
+        curl_list();
+
+        /**
+         * Appends the given string onto the list.
+         * @param value The string to append onto the list.
+         */
+        void append(std::string const& value);
+
+     private:
+        static void cleanup(curl_slist* list);
+    };
+
+    /**
+     * Resource for a cURL escaped string.
+     */
+    struct curl_escaped_string : util::scoped_resource<char const*>
+    {
+        /**
+         * Constructs a cURL escaped string.
+         * @param handle The cURL handle to use to perform the escape.
+         * @param str The string to escape.
+         */
+        curl_escaped_string(curl_handle const& handle, std::string const& str);
+
+     private:
+        static void cleanup(char const* str);
+    };
+
+    /**
+     * The exception for HTTP.
+     */
+    struct http_exception : std::runtime_error
+    {
+        /**
+         * Constructs an http_exception.
+         * @param message The exception message.
+         */
+        http_exception(std::string const &message) :
+            runtime_error(message)
+        {
+        }
+    };
+
+    /**
+     * The exception for HTTP requests.
+     */
+    struct http_request_exception : http_exception
+    {
+        /**
+         * Constructs an http_request_exception.
+         * @param req The HTTP request that caused the exception.
+         * @param message The exception message.
+         */
+        http_request_exception(request req, std::string const &message) :
```

Comment:
After talking with Maggie, I'm smoking crack here.  Ignore my suggestion about changing `message`.  The base type doesn't take the parameter by values, this is not what we should be doing.  I think I saw pass-by-const-ref, quickly glanced at the initializer list and saw the second item did a `std::move` and it invoked my wrong-o-meter without realizing that line was moving the request, not the message :/  Sorry about that!

With regards to the CA member functions, though, we should take those arguments by `const&` and not move or by-value and move.  I have no strong opinions on either as there are trade-offs for each.  Let's stick with taking them by `const&` and not moving for now.